t/lib/OpenQA/Test/FullstackUtils.pm

```diff
@@ -44,52 +43,42 @@ my $JOB_SETUP = {
 # equivalent, just saving some time
 $JOB_SETUP->{QEMU_NO_KVM} = '1' unless -r '/dev/kvm';
 
-sub job_setup {
-    my %override = @_;
+sub job_setup (%override) {
     my $new = {%$JOB_SETUP, %override};
     return join ' ', map { "$_=$new->{$_}" } sort keys %$new;
 }
 
-sub get_connect_args {
+sub get_connect_args () {
     my $mojoport = OpenQA::SeleniumTest::get_mojoport;
     return ["--apikey=1234567890ABCDEF", "--apisecret=1234567890ABCDEF", "--host=http://localhost:$mojoport"];
 }
 
-sub client_output {
-    my ($args) = @_;
-    my @connect_args = @{get_connect_args()};
-    open(my $client, "-|", "perl ./script/openqa-cli api @connect_args $args");
-    my $out;
-    while (<$client>) {
-        $out .= $_;
-    }
-    close($client);
-    return $out;
-}
+sub client_output ($args) { qx{perl ./script/openqa-cli api @{get_connect_args()} $args} }
 
-sub client_call {
-    my ($args, $expected_out, $desc) = @_;
+sub client_call ($args, $expected_out = undef, $desc = 'client_call') {
     my $out = client_output $args;
-    if ($expected_out) {
-        like($out, $expected_out, $desc) or die;
-    }
+    return undef unless $expected_out;
+    like($out, $expected_out, $desc) or die;
 }
 
-sub find_status_text { shift->find_element('#info_box .card-body')->get_text() }
+sub find_status_text ($driver) { $driver->find_element('#info_box .card-body')->get_text() }
 
-sub _fail_with_result_panel_contents {
-    my ($result_panel_contents, $msg) = @_;    # uncoverable statement
+# uncoverable statement
+sub _fail_with_result_panel_contents ($result_panel_contents, $msg) {
     diag("full result panel contents:\n$result_panel_contents");    # uncoverable statement
     ok(javascript_console_has_no_warnings_or_errors, 'No unexpected js warnings');    # uncoverable statement
     fail "Expected result not found";    # uncoverable statement
 }
 
-sub wait_for_result_panel {
-    my ($driver, $result_panel, $context, $fail_on_incomplete, $check_interval) = @_;
+sub wait_for_result_panel (
+    $driver, $result_panel,
+    $context = 'current job',
+    $fail_on_incomplete = undef,
+    $check_interval = 0.5
+  )
+{
     my $looking_for_result = $result_panel =~ qr/Result: /;
-    $context //= 'current job';
     my $msg = "Expected result for $context not found";
```

Comment:
CI tests show a warning about `$context` not defined, see https://app.circleci.com/pipelines/github/os-autoinst/openQA/8499/workflows/0d301351-4632-4b07-bf9f-e2036ff8c24b/jobs/80265?invite=true#step-110-99 , but https://github.com/os-autoinst/openQA/pull/4379/files#diff-523eaf01cc7dd1b0371426b1d310936bc8ae615a6b15cbd107ea0364ccbf8bdaR75 shows that there should be a default value. Did I do it wrong?

Reply:
Likely `undef` (or a variable with `undef` value) is passed explicitly by some invocation. Adding a signature with default value is *not* equivalent to assigning a "default" via `//=`.

Reply:
ok, using `undef` as default and setting the string accordingly.